Move address translation to opening new connection

wprzytula opened the pull request:
Motivations are described in the referenced issue.
Correctness of the changes is tested by the existing tests making use of address translation. 
Fixes: #640 

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~~[ ] I added relevant tests for new features and bug fixes.~~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I added appropriate `Fixes:` annotations to PR description.

piodul commented on scylla/src/transport/node.rs:
```diff
@@ -55,6 +57,72 @@ impl TimestampedAverage {
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy)]
+pub enum NodeAddr {
+    RealPeer(SocketAddr),       // fetched in Metadata with query_peers()
+    InitiallyKnown(SocketAddr), // provided by user in SessionConfig
+    FixedLocal(SocketAddr), // built from control connection's address in order to mitigate the bug described above
```
Please document each of the enum variants with a docstring.

piodul commented on scylla/src/transport/host_filter.rs:
```diff
@@ -54,7 +54,7 @@ impl AllowListHostFilter {
 impl HostFilter for AllowListHostFilter {
     fn accept(&self, peer: &Peer) -> bool {
-        self.allowed.contains(&peer.address)
+        self.allowed.contains(&peer.address.inner())
```
Hmm, so now we are consulting the host filter for all types of addresses (RealPeer, FixedLocal, InitiallyKnown). I think that we should only ask the hashmap about RealPeer addresses and return true for others:

- If the address is InitiallyKnown, the user put the address into the initial peers list, so they most likely want to connect to this node,
- If the address is FixedLocal, then the node either was originally an InitiallyKnown node, or a RealPeer node for which the host filter returned true.

What do you think?

wprzytula replied:
Does this preserve the HostFilter's behaviour regarding initially known nodes? Did the filter accept them all previously?

piodul replied:
Previously, the AllowListHostFilter was able to filter out initial nodes, but the driver would still be able to establish a control connection to one of them - if it happened, it would print a warning about it and switch the control connection to another node.

It's not really possible to preserve the semantics. Previously, AllowListHostFilter was always given a translated address, and now we don't do that - and we can't because the notion of a translated address makes less and less sense: node's address can change, multiple nodes can be behind a proxy, etc.

Maybe it's better to remove the AllowListHostFilter. It's not clear what its semantics should be. It's also quite simple for users to implement something like this if they need it and it will force them to think about how to handle different address types. Anyway, the main motivation for the host filter interface was introducing the DcHostFilter, which should still work.

piodul commented on scylla/src/transport/connection_pool.rs:
```diff
@@ -603,7 +605,9 @@ impl PoolRefiller {
 
                 evt = self.ready_connections.select_next_some(), if !self.ready_connections.is_empty() => {
-                    self.handle_ready_connection(evt);
+                    if let Ok(evt) = evt {
+                        self.handle_ready_connection(evt);
+                    }
```
You must call `self.handle_ready_connection` in case of failure. It updates the state of the pool as a reaction to the error.

`OpenedConnectionEvent` already holds a result, so you should put the translation error there (probably converted to a QueryError?).

wprzytula replied:
Do you find `QueryError` suitable for containing variants regarding _translating addresses on establishing connection to nodes_ ? For me, this is highly counterintuitive. I'm getting more and more convinced that it's high time we considered refactor of our errors.

wprzytula replied:
After all, there wasn't any query attempted yet, and the user gets a `QueryError`.

piodul replied:
> Do you find `QueryError` suitable for containing variants regarding _translating addresses on establishing connection to nodes_ ? For me, this is highly counterintuitive. I'm getting more and more convinced that it's high time we considered refactor of our errors.

I agree that our error types need redesign, but that is not something that we can fix in this PR.

In this case though it wouldn't hurt to introduce an OpenConnectionError or sth. The error is converted to a string [here](https://github.com/scylladb/scylla-rust-driver/blob/4964b4471c3d14bfea75e532b2e7c24cbd48cd65/scylla/src/transport/connection_pool.rs#L348-L356) (again, not ideal), but the new error won't appear in the API and will slightly improve the clarity of the internal code.

> After all, there wasn't any query attempted yet, and the user gets a `QueryError`.

To be fair, using QueryError to represent an error related to a failed connection is not perfect, either.

piodul commented on scylla/src/transport/connection_pool.rs:
```diff
@@ -907,45 +908,66 @@ impl PoolRefiller {
+                    Ok(addr)
+                }
+            }
+        };
```
I think this code could be pushed down to an even lower layer - maybe put the address translator to the connection config and then use it in `connection::open_conection`?

Moreover, you aren't considering the shard-aware port in this code, but you could.

wprzytula commented on scylla/src/transport/node.rs:
```diff
@@ -55,6 +57,72 @@ impl TimestampedAverage {
+        Self::RealPeer(*addr)
+    }
+}
+impl PartialEq for NodeAddr {
+    fn eq(&self, other: &Self) -> bool {
+        self.inner() == other.inner()
+    }
+}
+impl Eq for NodeAddr {}
+impl PartialOrd for NodeAddr {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.inner().partial_cmp(&other.inner())
+    }
+}
+impl Ord for NodeAddr {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.inner().cmp(&other.inner())
+    }
+}
+impl Hash for NodeAddr {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.inner().hash(state);
+    }
+}
```
I can't recall what made me choose this way. One idea: in case of no AddressTranslator provided, with current `known_peers` implementation, `Node`s will be recycled instead of being rebuilt if a control connection changes to another node: both previous and new control connection's `Node`s will be spared.

wprzytula commented on scylla/src/transport/node.rs:
```diff
@@ -55,6 +57,72 @@ impl TimestampedAverage {
+}
+impl NodeAddr {
+    pub(crate) fn inner(&self) -> SocketAddr {
+        match self {
+            NodeAddr::RealPeer(addr)
+            | NodeAddr::InitiallyKnown(addr)
+            | NodeAddr::FixedLocal(addr) => *addr,
+        }
```
`non_exhaustive` restricts some actions only beyond the defining crate.

piodul replied:
I see. Makes sense.

wprzytula commented on scylla/src/transport/cluster.rs:
```diff
@@ -60,7 +61,7 @@ pub struct Datacenter {
 #[derive(Clone)]
 pub struct ClusterData {
-    pub(crate) known_peers: HashMap<SocketAddr, Arc<Node>>, // Invariant: nonempty after Cluster::new()
+    pub(crate) known_peers: HashMap<NodeAddr, Arc<Node>>, // Invariant: nonempty after Cluster::new()
```
Do you propose releasing the constraint that, in order for a Node to be recycled, IPs need to match? I.e. now only Host ID, DC and rack must be the same to recycle a Node?

piodul replied:
> Do you propose releasing the constraint that, in order for a Node to be recycled, IPs need to match? I.e. now only Host ID, DC and rack must be the same to recycle a Node?

Hm, right, it would indeed be nice to update the IP field if it changes, so I think we should indeed create a new Node instance.

This leaves us with the problem of NodeConnectionPool being recreated. However, I realize now that this problem can be avoided:

1. Add `#[derive(Clone)]` for NodeConnectionPool. Most (if not all) fields are already cloneable, and if not we could put `refiller_handle` and `keepaliver_handle` into an Arc and it should work.
2. Make it possible to update the pool refiller with the new details about the node - in this case we only need to tell it about the new IP. We could do it with an ArcSwap or a tokio broadcast channel.

wprzytula replied:
So we drop the important invariant of NodeConnectionPool and PoolRefiller, which is constant IP. In fact we have already decided to do so because of serverless proxy hostname resolution.

wprzytula commented on scylla/src/transport/cluster.rs:
```diff
@@ -531,7 +532,7 @@ impl ClusterWorker {
         let cluster_data = self.cluster_data.load_full();
 
-        let node = match cluster_data.known_peers.get(&addr) {
+        let node = match cluster_data.known_peers.get(&NodeAddr::RealPeer(addr)) {
```
I've moved to the default `Eq` etc. impls.

wprzytula commented on scylla/src/transport/topology.rs:
```diff
@@ -61,7 +61,6 @@ pub struct Peer {
 
 #[non_exhaustive] // <- so that we can add more fields in a backwards-compatible way
 pub struct UntranslatedPeer {
-    pub host_id: Uuid,
```
I've kept it back and cherry picked the commit introducing `host_id` into ConnectionPool from serverless PR.

wprzytula commented:
v2:
- addressed all comments,
- considerable change in the last commits: when a `node_addr` changes, a new Node is created but the pool from the old Node is preserved and updated with the new address.

wprzytula commented:
Rebased on main.

wprzytula commented:
v3:
- refactored topology, connection_pool and connection layers so that they employ new `UntranslatedEndpoint` enum to carry the data necessary for opening new connections, including information whether to perform address translation or not.

wprzytula commented:
@piodul @havaker @cvybhu Ready for review.

piodul commented on scylla/src/transport/topology.rs:
```diff
@@ -61,6 +61,66 @@ pub struct Peer {
+    pub rack: Option<String>,
+}
+
+impl From<Peer> for PeerEndpoint {
+    fn from(peer: Peer) -> Self {
+        Self {
+            host_id: peer.host_id,
+            address: peer.address,
+            datacenter: peer.datacenter,
+            rack: peer.rack,
+        }
+    }
+}
```
Do we need to expose this conversion? Is it used anywhere? I'm not sure whether it is useful to convert a Peer by value into a PeerEndpoint.

I'd rather not introduce such impls unnecessarily as they are a part of the API.

wprzytula replied:
Right, it's currently unused.

piodul commented on scylla/src/transport/topology.rs:
```diff
@@ -267,7 +268,7 @@ impl Metadata {
 
                 Peer {
-                    address: endpoint.address(),
+                    address: NodeAddr::RealPeer(endpoint.address()),
```
I don't think we can treat an initial address as a `RealPeer`. This function will be called with contact point addresses because it is only called on first metadata fetch, so those addresses shouldn't be translated. If address translation is enabled then, in case of a failed first metadata fetch, the session will not work anymore.

The semantics of `FixedLocal` are more appropriate here, though the name isn't. I propose to change the variant names of `NodeAddr` to something more generic - something that only refers to whether the contained address should be translated or not.

By the way, it's too easy to construct a wrong `NodeAddr` from an `UntranslatedEndpoint` via the `.address()` method.

wprzytula replied:
I adjusted the names and usage as suggested. However, I can't see any sensible solution the to problem of creating erroneuously wrong `NodeAddr` via `.address()`.

wprzytula replied:
Do you have any ideas @piodul?

wprzytula replied:
I've rewritten the `address()` method to return `NodeAddr` always. In case of a `ContactPoint`, a `NodeAddr::Untranslatable` is returned.

piodul commented on scylla/src/transport/topology.rs:
```diff
@@ -463,7 +464,7 @@ impl MetadataReader {
             .peers
             .iter()
-            .find(|peer| peer.address == self.control_connection_endpoint.address());
+            .find(|peer| peer.address.into_inner() == self.control_connection_endpoint.address());
```
I think this is wrong. We shouldn't compare SocketAddrs but rather NodeAddrs. We cannot consider `RealPeer(x)` to be equal to `FixedLocal(x)` for any `x`.

wprzytula commented on scylla/src/transport/connection_pool.rs:
```diff
@@ -203,6 +204,13 @@ impl NodeConnectionPool {
         }
     }
 
+    pub(crate) fn update_node_addr(&self, new_node_addr: NodeAddr) {
```
OK, it seems that any use case of `update_node_addr` should have access to a whole `Peer`, which imples `PeerEndpoint` as well. We could do so then. However, what bothers me is:
- (minor thing) pointless clones of DC and Rack ,
- (major thing) Pool's API is now not immune to alterations of node parameters other than only address. From the perspective of the Pool, it is now possible that different rack, host ID or DC are passed.

piodul replied:
> * (minor thing) pointless clones of DC and Rack ,

This happens very infrequently so it's not a big problem.

> * (major thing) Pool's API is now not immune to alterations of node parameters other than only address. From the perspective of the Pool, it is now possible that different rack, host ID or DC are passed.

Why is it a major problem?

wprzytula replied:
> Why is it a major problem?

Alterations (especially silent) of the above stated node properties are not something we designed ConnectionPool for.

wprzytula replied:
But OK, the Pool won't break in such a case, and we don't intend to cause such situation anyway, so let's enable complete endpoint updates.
